### PR TITLE
feat: add support for custom workspacetypes for newly created organizations

### PR DESCRIPTION
**Pull request by kekcleader**
#16 

## **Problem Solved**  
- Admission errors when creating org workspaces under `root:orgs` due to parent/extension mismatches  
- Missing extension relationships between custom and base `WorkspaceTypes`  
- Inflexible org workspace placement (hardcoded cluster locations)  

### **1. WorkspaceType Extension Relationships**  
- Added `Extend.With` field to both custom org and account `WorkspaceTypes`  
- Custom org type now properly extends base `"org"` type:  
  ```yaml
  Extend.With: 
    - Name: "org"
      Path: "root"
  ```  
- Custom account type now properly extends base `"account"` type:  
  ```yaml
  Extend.With: 
    - Name: "account"
      Path: "root"
  ```  
- Resolves admission controller validation by establishing proper inheritance hierarchy  

### **2. Configurable Org Workspace Cluster**  
- New configuration field: `kcp-org-workspace-cluster` (environment variable)  
- Flexible org workspace placement: specify custom cluster for org workspace creation  
- Backward compatible: defaults to current cluster behavior when not specified  
- Environment-based configuration using `mapstructure` tags  

### **3. Enhanced Parent-Child Constraints**  
- Custom org `WorkspaceType`: strictly allows only `orgs@root` as parent  
- Custom account `WorkspaceType`: allows parent types `custom-org` (current cluster) and self  
- Proper path handling: updated logic for base type updates with configurable clusters  
- Cross-cluster compatibility: maintains proper cluster context for workspace operations  

### **4. Root Client Derivation**  
- Enhanced root client creation: strips `/services` and `/clusters` for base type lookups  
- KCP-aware `RESTMapper`: uses manager’s `RESTMapper` for proper cluster routing  
- Fallback mechanisms: graceful handling when root client unavailable  

### **5. Workspace Creation Logic**  
- Org workspaces: created under `root:orgs` using root client when configured  
- WorkspaceType references: use explicit path and name for proper resolution  
- Cluster-aware operations: context switching for cross-cluster workspace creation  

### **6. Comprehensive Test Coverage**  
Added/updated tests:  
- Extension relationship validation (`Extend.With` fields set correctly)  
- Root host derivation and fallback (client creation edge cases)  
- Error injection: `Get` / `Create` / `Patch` failure scenarios  
- Missing cluster context: graceful handling without panic  
- Base type not found: continues operation without base types  
- Parent constraint validation: ensures proper parent-child relationships  
- `DefaultAPIBindings` inheritance: verifies spec copying and fallback logic  
- Configurable cluster testing: validates `OrgWorkspaceCluster` functionality  

### **Configuration**  
```go
type KcpConfig struct {
    // ... existing fields ...
    OrgWorkspaceCluster string `mapstructure:"kcp-org-workspace-cluster" default:""`
}
```

### **Extension Setup**  
```go
// Custom org extends base org
customOrg.Spec.Extend = kcptenancyv1alpha.WorkspaceTypeExtension{
    With: []kcptenancyv1alpha.WorkspaceTypeReference{
        {Name: "org", Path: "root"},
    },
}

// Custom account extends base account
customAcc.Spec.Extend = kcptenancyv1alpha.WorkspaceTypeExtension{
    With: []kcptenancyv1alpha.WorkspaceTypeReference{
        {Name: "account", Path: "root"},
    },
}
```

### **Test Validation**  
```go
len(wtOrg.Spec.Extend.With) == 1 &&
string(wtOrg.Spec.Extend.With[0].Name) == "org" &&
len(wtAcc.Spec.Extend.With) == 1 &&
string(wtAcc.Spec.Extend.With[0].Name) == "account"
```

## **Results**  
- Admission errors eliminated: org workspaces can now be created under `root:orgs`  
- Proper inheritance: custom types correctly extend base types  
- Flexible deployment: configurable org workspace cluster location  
- All tests passing: comprehensive test coverage validates functionality  
- Backward compatible: existing behavior preserved when config not specified  

## **Testing Status**  
- Unit tests (subroutines): **All passing** 
- Integration tests (controller): **All passing**
- API tests (v1alpha1): **All passing**  
- Full test suite: **All passing with `task test`**  

## Summary by CodeRabbit

* **New Features**
  * Auto-provisions WorkspaceTypes for orgs/accounts with inheritance, auth wiring, and Kubernetes-safe naming helpers.

* **Configuration**
  * New toggles and delays for workspace-type behavior; root-host and org-cluster overrides; IMAGE_REPO added to Taskfile.

* **Reliability**
  * Fallback to shared client when root client unavailable; optional relaxed handling of Forbidden workspace creation with configurable requeue.

* **Metrics**
  * Workspace creation forbidden metric added and registered (with label sanitization).

* **Refactor**
  * Switched to standard controller-runtime manager.

* **Tests**
  * Expanded coverage for naming, provisioning, virtual/root-host derivation, forbidden/error paths, and reconcilers.

* **Documentation**
  * Added end-to-end testing guide for custom WorkspaceTypes and KCP setup.

* **Dependencies/Tooling**
  * Updated dependency replacements/versions and more robust local-binary setup in Taskfile.

### CODE REVIEW

File: internal/controller/account_controller.go [hunk 1]
```diff
@@ -45,6 +49,37 @@ type AccountReconciler struct {
...
+			if idx := strings.Index(base, "/services/"); idx != -1 {
+				base = base[:idx]
+			}
+			// If we're pointing at a specific cluster, strip it to get the server base
+			if idx := strings.Index(base, "/clusters/"); idx != -1 {
+				base = base[:idx]
+			}
+			rootHost = strings.TrimRight(base, "/") + "/clusters/root"
+		}
+		if rootHost != "" {
```
[the remark this answers is not in the document]
Comment (kekcleader) [reply internal/controller/account_controller.go]: Lint will give an error in this case. Maybe it is better to do this:

If `rootHost` is empty → derive it from manager config
In any case → create client with `rootHost` (either from config or derived)

**Review by nexus49 — CHANGES_REQUESTED** [review]
I've reviewed the PR and the solution does not address the task in a way that i feel we can correct it by putting individual PR comments. It introduces 3000 lines of code change which is not appropriate to the task at hand. Lets have a meeting to discuss the solution in an alternative PR that i'll prepare.
